core/lib/include/assert.h

```diff
@@ -67,7 +67,7 @@ extern "C" {
 #endif /*__NORETURN*/
 
 #ifdef NDEBUG
-#define assert(ignore)((void)0)
+#define assert(cond) if (!(cond)) __builtin_unreachable()
```

Comment:
https://godbolt.org/z/sozoehc6s doesn't compile with this implementation of `assert()` :) That's an interesting puzzle to solve. *summoning* @kfessel

Reply:
Is mixing asserts with other statements on the same line really something we do (or should do)?

Reply:
Also, what about all these bugs:

```C
assert(spi_acquire(bus, ...) == SPI_OK);
```

that would no longer be bugs with this? :smiling_imp: :stuck_out_tongue_winking_eye:

Reply:
```C
#define assert(cond) __builtin_expect((cond), 0)
```

Seems to at least compile for all non-bug use cases of `assert()` correctly; but it would still mitigate bugs. And that certainly is a bad idea to fix the broken-by-design `assert()` API contract ;) (At least that was the conclusion last time such a fix was PRed.)

Reply:
oddly enough that one gives me a 

```
/home/benpicco/dev/RIOT/core/lib/include/assert.h:70:22: error: statement with no effect [-Werror=unused-value]
   70 | #define assert(cond) __builtin_expect(!(cond), 0)
      |                      ^~~~~~~~~~~~~~~~~~~~~~~~~~~~
/home/benpicco/dev/RIOT/core/lib/include/cib.h:61:5: note: in expansion of macro 'assert'
   61 |     assert(!(size & (size - 1)));
      |     ^~~~~~
```

Clang also has a `__builtin_assume()` but GCC does not.

Reply:
> Is mixing asserts with other statements on the same line really something we do (or should do)?

Definitely not! We do have some really creative code in some of our packages and vendor headers, though :) And `assert()` is used commonly enough that there may be some fallout if it is not mixing well with creative code.

Reply:
let's see if this is actually an issue

Reply:
turns out lz4 is doing *exactly that*

```C
    LZ4_streamDecode_t_internal* const lz4sd =
        (assert(LZ4_streamDecode!=NULL), &LZ4_streamDecode->internal_donotuse);
```

Reply:
```
#define assert(cond) (cond)?((void)0):__builtin_unreachable()
```